pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBaseTxnBufferSnapshotService.java

```diff
@@ -18,69 +18,168 @@
+                return null;
+            });
+        }
+
+        public CompletableFuture<Writer<TransactionBufferSnapshot>> getFuture() {
+            if (future == null) {
+                initWriterFuture();
+            }
+            return future;
+        }
+
+        @Override
+        protected void deallocate() {
+            ReferenceCountedWriter referenceCountedWriter = service.writerFutureMap.remove(namespaceName);
+            if (referenceCountedWriter != null && referenceCountedWriter.getFuture() != null) {
+                service.pendingCloseWriterList.add(referenceCountedWriter.getFuture());
+                service.closePendingCloseWriter();
+            }
+        }
+
+        @Override
+        public ReferenceCounted touch(Object o) {
+            return this;
+        }
+
+    }
+
+    public SystemTopicBaseTxnBufferSnapshotService(PulsarClient client,
+                                                   ScheduledExecutorService scheduledExecutorService) {
         this.namespaceEventsSystemTopicFactory = new NamespaceEventsSystemTopicFactory(client);
         this.clients = new ConcurrentHashMap<>();
+        this.scheduledExecutorService = scheduledExecutorService;
+        this.writerFutureMap = new ConcurrentHashMap<>();
+        this.pendingCloseWriterList = new LinkedList<>();
     }
 
     @Override
     public CompletableFuture<Writer<TransactionBufferSnapshot>> createWriter(TopicName topicName) {
-        return getTransactionBufferSystemTopicClient(topicName).thenCompose(SystemTopicClient::newWriterAsync);
-    }
-
-    private CompletableFuture<SystemTopicClient<TransactionBufferSnapshot>> getTransactionBufferSystemTopicClient(
-            TopicName topicName) {
-        TopicName systemTopicName = NamespaceEventsSystemTopicFactory
-                .getSystemTopicName(topicName.getNamespaceObject(), EventType.TRANSACTION_BUFFER_SNAPSHOT);
-        if (systemTopicName == null) {
+        if (topicName == null) {
             return FutureUtil.failedFuture(
-                    new InvalidTopicNameException("Can't create SystemTopicBaseTxnBufferSnapshotService, "
-                            + "because the topicName is null!"));
+                    new PulsarClientException.InvalidTopicNameException(
+                            "Can't create SystemTopicBaseTxnBufferSnapshotService, because the topicName is null!"));
         }
-        return CompletableFuture.completedFuture(clients.computeIfAbsent(systemTopicName,
+        return getTransactionBufferSystemTopicClient(topicName.getNamespaceObject()).newWriterAsync();
+    }
+
+    @Override
+    public ReferenceCountedWriter createReferenceWriter(NamespaceName namespaceName) {
+        return writerFutureMap.compute(namespaceName, (ns, writerFuture) -> {
+            if (writerFuture == null) {
+                return new ReferenceCountedWriter(namespaceName, this);
```

Comment:
Seems not safe here.

Reply:
This is a ConcurrentHashMap, it should be thread safe.

Reply:
yes.